Task 3.2

Mitoschka opened the pull request.

Mitoschka committed:
Add Task 3.2 (No tests)

Mitoschka committed:
Update .yml

Mitoschka committed:
Task 3.2

Mitoschka committed:
Update Task 3.2

Mitoschka committed:
Merge branch 'master' into Task-3.2

yurii-litvinov commented on Task 3.2/Task 3.2.Test/FirstHash.Test.cs:
```diff
@@ -0,0 +1,36 @@
+        public void AddElementToHashTableAndDeleteElementOfHashTableShouldWork()
+        {
+            IMyHash myHash = new FirstHash();
+            HashTable hashTable = new HashTable(myHash);
```
Лучше вынести в Setup()

yurii-litvinov commented on Task 3.2/Task 3.2.Test/Hash.Test.cs:
```diff
@@ -0,0 +1,74 @@
+            HashTable hashTable = new HashTable(myHash);
+
+            hashTable.AddElementToHashTable("abc");
+            myHash = new SecondHash();
```
Чего Вы хотели добиться этой строчкой?

yurii-litvinov commented on Task 3.2/Task 3.2.Test/Hash.Test.cs:
```diff
@@ -0,0 +1,74 @@
+            myHash = new ThirdHash();
+            hashTable.AddElementToHashTable("abc");
+            myHash = new FirstHash();
+            myHash = new SecondHash();
```
Оу

yurii-litvinov commented on Task 3.2/Task 3.2.Test/ThirdHash.Test.cs:
```diff
@@ -0,0 +1,36 @@
+
+            Assert.IsFalse(hashTable.IsContainInHashTable("abc"));
+        }
+    }
```
Опять-таки, копипаст в тестах

yurii-litvinov commented on Task 3.2/Task 3.2/FirstHash.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public class FirstHash : IMyHash
+    {
+        const int capacity = 100;
```
А зачем хеш-функции какой-то capacity, который никак не связан с размером хеш-таблицы, потому что это два разных класса?

yurii-litvinov commented on Task 3.2/Task 3.2/IMyHash.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Task_3._2
+{
+    public interface IMyHash
```
Надо комментарии, и к самому интерфейсу, и к каждому его методу

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <param name="value">Value to add</param>
+        public void AddElementToHashTable(string value)
+        {
+            int index = Hash.HashFunction(value);
```
Тут мы не можем никак гарантировать, что HashFunction вернёт нам число в диапазоне от 0 до hashTable.Length - 1, потому что хеш-функцию мы принимаем извне и её может написать любой школьник. Так что тут надо предполагать, что она может вернуть что угодно и приводить в нужный диапазон самим.

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+        /// Pop the element
+        /// </summary>
+        /// <returns> element </returns>
+        public string PopElement()
```
А что это за метод и зачем он хеш-таблице?

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <summary>
+        /// Add element to hash table
+        /// </summary>
+        public void AddElementToHashTable()
```
Метод, который, судя по названию, добавляет элемент в хеш-таблицу, но не принимает элемент, выглядит очень безумно

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <param name="value">Value to be deleted</param>
+        public void DeleteElementOfHashTable(string value)
+        {
+            int index = Hash.HashFunction(value);
```
А зачем считать хеш-функцию, если мы ещё не знаем, есть ли вообще такой элемент в хеш-таблице?

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+            if (IsContainInHashTable(value) == true)
+            {
+                hashTable[index].DeleteElement(value);
+                Console.WriteLine($"\"{value}\" было удалено" + "\n");
```
Нельзя ничего писать на консоль

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,109 @@
+        public void DeleteElementOfHashTable(string value)
+        {
+            int index = Hash.HashFunction(value);
+            if (IsContainInHashTable(value) == true)
```
Нельзя сравниваться с булевыми константами :)

yurii-litvinov commented on Task 3.2/Task 3.2/SecondHash.cs:
```diff
@@ -0,0 +1,38 @@
+        /// </summary>
+        /// <param name="degree">Degree of received number</param>
+        /// <returns>Returns the power of a number</returns>
+        public int PowerOfThree(int degree)
```
Зачем она public?

Mitoschka committed:
1

Mitoschka committed:
Update

Mitoschka committed:
Update

Mitoschka committed:
Add more tests and corrected comments

Mitoschka committed:
Merge branch 'master' into Task-3.2

yurii-litvinov commented on Task 3.2/Task 3.2/FirstHash.cs:
```diff
@@ -0,0 +1,21 @@
+﻿namespace Task_3._2
+{
+    public class FirstHash : IMyHash
```
Надо комментарий и к самим классам тоже

yurii-litvinov commented on Task 3.2/Task 3.2/IMyHash.cs:
```diff
@@ -0,0 +1,15 @@
+        /// added value to HashFunction
+        /// </summary>
+        /// <param name="value"> value added to HashFunction</param>
+        /// <returns></returns>
```
А вот пустые комментарии не нужны

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <summary>
+        /// Creat HashTable
+        /// </summary>
+        /// <param name="myHash"> instance of classe, its auction </param>
```
classe? auction?

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,86 @@
+    public class HashTable
+    {
+        /// <summary>
+        /// Creat HashTable
```
Make HashTable creat again? :) Надо внимательней в комментариях

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <param name="myHash"> changed hash </param>
+        public void ChangeHash(IMyHash myHash)
+        {
+            MyList list = new MyList();
```
var

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <param name="value">Value to add</param>
+        public void AddElementToHashTable(string value)
+        {
+            int index = Hash.HashFunction(value) % capacity;
```
HashFunction может вернуть отрицательное число, тогда всё сломается

yurii-litvinov replied:
Напоминаю

yurii-litvinov commented on Task 3.2/Task 3.2/MyList.cs:
```diff
@@ -0,0 +1,156 @@
+{
+    public class MyList
+    {
+        class ListElement
```
private

yurii-litvinov commented on Task 3.2/Task 3.2/Program.cs:
```diff
@@ -0,0 +1,144 @@
+            "'3' - Проверить значение на принадлежность \n" +
+            "'4' - Изменить хэш - функцию \n" +
+            "'5' - Вызвать меню команд \n");
+        }
```
Пустой строки после метода не хватает

yurii-litvinov commented on Task 3.2/Task 3.2/Program.cs:
```diff
@@ -0,0 +1,144 @@
+{
+    public class Program
+    {
+        static void PrintMenu()
```
private

Mitoschka committed:
Update task

Mitoschka committed:
Merge branch 'master' into Task-3.2

Mitoschka committed:
Update Task

yurii-litvinov commented on Task 3.2/Task 3.2.Test/Hash.Tests.cs:
```diff
@@ -0,0 +1,55 @@
+
+        }
+
+        class DivideCases : IEnumerable
```
DivideCases --- "случаи деления"? Я бы рекомендовал более осмысленное имя, например, Hashes

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace Task_3._2
+{
+    public class HashTable
```
Ну, а тут поленились комментарий написать? :)

yurii-litvinov replied:
Напоминаю

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+            int index = Hash.HashFunction(value) % capacity;
+            if (index < 0)
+            {
+                throw new System.IndexOutOfRangeException("negative array index value");
```
Нет, так Вы сделали, чтобы оно сломалось управляемо, но оно же всё равно ломается. Я как пользователь пишу хеш-функцию, передаю Вашей хеш-таблице, и тут, внезапно, получаю IndexOutOfRangeException при добавлении элемента. Иногда. И думаю: "какой индекс? почему OutOfRange? я просто хочу добавить строку в множество". Нет конечно, в качестве хеш-функции должно быть можно использовать стандартный Object.GetHashCode

yurii-litvinov commented on Task 3.2/Task 3.2/MyHash.cs:
```diff
@@ -0,0 +1,90 @@
+        /// <returns>Returns index position</returns>
+        public bool IsContainInHashTable(string value)
+        {
+            int index = Hash.HashFunction(value) % capacity;
```
А тут вообще забили на проблему :)

yurii-litvinov commented on Task 3.2/Task 3.2/MyList.cs:
```diff
@@ -0,0 +1,168 @@
+
+namespace Task_3._2
+{
+    public class MyList
```
Список надо бы тоже обновить после победы над задачей 2.1. И таки написать комментарий, для этой задачи они обязательны

yurii-litvinov replied:
Напоминаю

yurii-litvinov commented on Task 3.2/Task 3.2/Program.cs:
```diff
@@ -0,0 +1,145 @@
+        static void Main()
+        {
+            IMyHash myHash;
+            myHash = new ThirdHash();
```
Это можно на одной строке писать

Mitoschka committed:
Update test of task and update task

Update task:
1) MyHash.cs : added a condition for checking the index whether it is negative or whether it goes beyond the hash of the table.
2) MyList.cs: added invariant for head and tail.
3) Program.cs: shortened two lines, writing them in one.
Update test of task:
1)MyList.Tests.cs: changed the exception to another, more suitable.
2)Hash.Teste.cs:  changed the class name "DivideCases" to "Hashes".

Mitoschka committed:
Update Task

removed the unnecessary cycle

Mitoschka committed:
Add new XML

Mitoschka committed:
Merge branch 'master' into Task-3.2

yurii-litvinov commented on Task 3.2/Task 3.2.Test/MyList.Tests.cs:
```diff
@@ -0,0 +1,50 @@
+        [Test]
+        public void CheckForEmptyList()
+        {
+            Assert.Throws<System.NullReferenceException>(() => myList.DeleteElement(" "));
```
Не, NullReferenceException не должно быть в принципе. Всегда, когда бросается NullReferenceException, это должно означать ошибку в самом коде или в деплое приложения, его никогда-никогда не надо обрабатывать.

Mitoschka committed:
Fix Exception and XML

yurii-litvinov reviewed (approved):
Good enough, зачтена

Mitoschka committed:
Merge branch 'master' into Task-3.2
